File: lua/vimade/style/value/animate.lua

local math = require('math')
local EASE = require('vimade.style.value.ease')
local CONDITION = require('vimade.style.value.condition')
local DIRECTION = require('vimade.style.value.direction')
local ANIMATOR = require('vimade.animator')
local COLOR_UTIL = require('vimade.util.color')
local TYPE = require('vimade.util.type')
local M = {}
local GLOBALS

local MATH_FLOOR = math.floor
local MATH_MIN = math.min
local MATH_MAX = math.max
local INTERPOLATE_FLOAT = COLOR_UTIL.interpolateFloat
local INTERPOLATE_RGB = COLOR_UTIL.interpolateRgb

M.__init = function (args)
  GLOBALS = args.GLOBALS
end

local DEFAULT_DURATION = 300
local DEFAULT_DELAY = 0
local DEFAULT_EASE = EASE.OUT_QUART
local DEFAULT_DIRECTION = DIRECTION.OUT

local ids = 0
local get_next_id = function ()
  ids = ids + 1
  return ids
end

M.EASE = EASE
M.DIRECTION = DIRECTION 

--@param config {
--  @required to = number | function -> number
--  @optional id: string | number | function -> string = 0 -- used to share state between values that might cross over between filters, exclusions, and other rules
--  @optional start: number | function -> number = 0
--  @optional duration: number | function -> number = 300
--  @optional delay: number | function -> number = 0
--  @optional ease: EASE | function -> EASE = EASE.OUT_QUART
--  @optional direction: DIRECTION | function -> DIRECTION 
--}
M.Number = function (config)
  config = TYPE.shallow_copy(config)
  config.interpolate = INTERPOLATE_FLOAT
  config.compare = function (value, last)
    return value == last
  end
  return M.Animate(config)
end

M.Rgb = function(config)
  config = TYPE.shallow_copy(config)
  config.interpolate = INTERPOLATE_RGB
  config.compare = TYPE.shallow_compare
  return M.Animate(config)
end

M.Invert = function(config)
  config = TYPE.shallow_copy(config)
  config.interpolate = function (to, start, pct_time, style, state)
    local result = {}
    to = to or {}
    start = start or {}
    result.fg = INTERPOLATE_FLOAT(to.fg or 0, start.fg or 0, pct_time)
    result.bg = INTERPOLATE_FLOAT(to.bg or 0, start.bg or 0, pct_time)
    result.sp = INTERPOLATE_FLOAT(to.sp or 0, start.sp or 0, pct_time)
    return result
  end
  config.compare = TYPE.deep_compare
  return M.Animate(config)
end

-- interpolate the entire available tint object
M.Tint = function(config)
  config = TYPE.shallow_copy(config)

  config.interpolate = function (to, start, pct_time)
    if not start and not to then
      return nil
    end
    to = to or {}
    start = start or {}
    local result = {}
    for key, value in pairs(to) do
      if start[key] == nil then
        start[key] = {rgb = value.rgb, intensity = 0}
      end
    end
    for key, value in pairs(start) do
      if to[key] == nil then
        to[key] = {rgb = value.rgb, intensity = 0}
      end
    end
    for key, value in pairs(to) do
      result[key] = {
        rgb = INTERPOLATE_RGB(value.rgb, start[key].rgb, pct_time),
        intensity = INTERPOLATE_FLOAT(value.intensity, start[key].intensity, pct_time),
      }
    end
    return result
  end
  config.compare = TYPE.deep_compare
  return M.Animate(config)
end

--@param config {
--  @required to = number | function -> number
--  @required interpolate = function(pct_time, start_value, to_value) -> value
--  @optional id: string | number | function -> string = 0 -- used to share state between values that might cross over between filters, exclusions, and other rules
--  @optional start: number | function -> number = 0
--  @optional duration: number | function -> number = 300
--  @optional delay: number | function -> number = 0
--  @optional ease: EASE | function -> EASE = EASE.OUT_QUART
--  @optional direction: DIRECTION | function -> DIRECTION 
--}
M.Animate = function (config)
  -- TODO abstract this part (to use with all animations)
  local _custom_id = config.id
  local _id = _custom_id or get_next_id()
  local _interpolate = config.interpolate
  local _start = config.start
  local _to = config.to
  local _duration = config.duration or DEFAULT_DURATION
  local _delay = config.delay or DEFAULT_DELAY
  local _ease = config.ease or DEFAULT_EASE
  local _direction = config.direction or DEFAULT_DIRECTION
  local _reset = config.reset
  -- For direction IN_OUT the to & start values swap around, so
  -- the reset value also needs to respect that
  local _reset_swap = _direction == DIRECTION.IN_OUT
  if _reset == nil then
    _reset = true
  end
  local _compare = config.compare or nil
  return function(style, state)
    local id
    local win = style.win.area_owner or style.win
    if _custom_id then
      state = state.custom
      if type(_custom_id) == 'function' then
        id = _custom_id(style, state)
      else
        id = _custom_id
      end
    else
      state = state.animations
      id = _id
    end
    if not state[id] then
      state[id] = {}
    end
    state = state[id]
    local to = style.resolve(_to, state)
    local start = style.resolve(_start, state)
    local compare = _compare
    if type(compare) == 'function' then
      compare = compare(to, state.last_to)
    end
    if compare == false then
      state.change_timestamp = GLOBALS.now
    end
    state.last_to = to
    local delay = _delay
    if type(delay) == 'function' then
      delay = delay(style, state)
    end
    local duration = _duration
    if type(duration) == 'function' then
      duration = duration(style, state)
    end
    local direction = _direction
    if type(direction) == 'function' then
      direction = direction(style, state)
    end
    local reset = _reset
    if type(reset) == 'function' then
      reset = reset(style, state)
    end
    -- Deterministically round to the nearest 16ms, this seems to just look better more
    -- consistently. 32ms would be better for bucketing, we need to revisit this.
    -- TODO make this configurable
    local time = MATH_FLOOR((GLOBALS.now - (MATH_MAX(win.timestamps.nc, state.change_timestamp or 0) + delay)) / 16) * 16
    -- TODO this logic is nc specific and should be abstracted
    if (direction == DIRECTION.OUT and style._condition == CONDITION.ACTIVE and win.nc == false)
      or (direction == DIRECTION.IN and (style._condition == CONDITION.INACTIVE or style._condition == CONDITION.INACTIVE_OR_FOCUS) and win.nc == true) then
       state.value = to
       state.start = to
       -- TODO when setting animating it should apply to the style and all parents
       style._animating = false
       return to
    elseif (direction == DIRECTION.OUT and style._condition == CONDITION.ACTIVE and win.nc == true)
      or (direction == DIRECTION.IN and (style._condition == CONDITION.INACTIVE or style._condition == CONDITION.INACTIVE_OR_FOCUS) and win.nc == false) then
        local swp = start
        start = to
        to = swp
    elseif (direction == DIRECTION.OUT and (style._condition == CONDITION.INACTIVE or style._condition == CONDITION.INACTIVE_OR_FOCUS) and win.nc == false)
      or (direction == DIRECTION.IN and style._condition == CONDITION.ACTIVE and win.nc == true) then
      state.value = start
      state.start = start
      style._animating = false
      return start
    elseif (direction == DIRECTION.OUT and style._condition == CONDITION.INACTIVE and win.nc == true)
      or (direction == DIRECTION.IN and style._condition == CONDITION.ACTIVE and win.nc == false) then
      -- pass
    end

    if state.value == nil then
      state.value = start
      state.start = start
    end
    if time <= 0 then
      if reset == true then
        state.start = (_reset_swap and state.to) or start
        state.value = (_reset_swap and state.to) or start
      else
        state.start = state.value
      end
      style._animating = true
      ANIMATOR.schedule(style)
      return state.start
    end
    state.to = to
    state.start = start
    if time >= duration then
      state.value = to
      style._animating = false
      return to
    end
    local elapsed = time / duration
    -- Round elapsed time to the closest 1000th (helps with deterministic bucketing).
    -- the output used from this is for fading and color manipulation, so this level
    -- of granularity would not even be user perceivable.
    elapsed = MATH_FLOOR(MATH_MIN(MATH_MAX(_ease(elapsed), 0), 1) * 1000 + 0.5) / 1000
    state.value = _interpolate(to, state.start, elapsed, style, state)
    style._animating = true
    ANIMATOR.schedule(style)
    return state.value
  end
end

return M

File: lib/vimade/style/value/animate.py
import sys
M = sys.modules[__name__]

from vimade import animator as ANIMATOR
from vimade.style.value import ease as EASE
from vimade.style.value import condition as CONDITION
from vimade.style.value import direction as DIRECTION
from vimade.style import invert as INVERT
from vimade.util import type as TYPE
from vimade.util import color as COLOR_UTIL

DEFAULT_DURATION = 300
DEFAULT_DELAY = 0
DEFAULT_EASE = EASE.OUT_QUART
DEFAULT_DIRECTION = DIRECTION.OUT

def __init(args):
  global GLOBALS
  GLOBALS = args['GLOBALS']
M.__init = __init

ids = 0
def get_next_id():
  global ids
  ids = ids + 1
  return ids

M.EASE = EASE
M.DIRECTION = DIRECTION 

def _compare_eq(value, last):
  return value == last

def Number(**kwargs):
  def interpolate(to, start, pct_time, style, state):
    return COLOR_UTIL.interpolateFloat(to, start, pct_time)
  kwargs['interpolate'] = interpolate
  kwargs['compare'] = _compare_eq
  return M.Animate(**kwargs)

def Rgb(**kwargs):
  def interpolate(to, start, pct_time, style, state):
    return COLOR_UTIL.interpolateRgb(to, start, pct_time)
  kwargs['interpolate'] = interpolate
  kwargs['compare'] = TYPE.shallow_compare
  return M.Animate(**kwargs)

def Invert(**kwargs):
  def interpolate(to, start, pct_time, style, state):
    result = {}
    to = to or {}
    start = start or {}
    result['fg'] = COLOR_UTIL.interpolateFloat(to.get('fg', 0), start.get('fg', 0), pct_time)
    result['bg'] = COLOR_UTIL.interpolateFloat(to.get('bg', 0), start.get('bg', 0), pct_time)
    result['sp'] = COLOR_UTIL.interpolateFloat(to.get('sp', 0), start.get('sp', 0), pct_time)
    return result
  kwargs['interpolate'] = interpolate
  kwargs['compare'] = TYPE.deep_compare
  return M.Animate(**kwargs)

def Tint(**kwargs):
  def interpolate(to, start, pct_time, style, state):
    if not start and not to:
      return None
    to = to or {}
    start = start or {}
    result = {}
    for key, value in to.items():
      if not key in start:
        start[key] = {'rgb': value['rgb'], 'intensity': 0}
    for key, value in start.items():
      if not key in to:
        to[key] = {'rgb': value['rgb'], 'intensity': 0}
    for key, value in to.items():
      result[key] = {
        'rgb': COLOR_UTIL.interpolateRgb(value['rgb'], start[key]['rgb'],  pct_time),
        'intensity': COLOR_UTIL.interpolateFloat(value['intensity'], start[key]['intensity'], pct_time),
      }
    return result
  kwargs['interpolate'] = interpolate
  kwargs['compare'] = TYPE.deep_compare
  return M.Animate(**kwargs)

# @param kwargs {
#  @required to = number | function -> number
#  @required interpolate = function(pct_time, start_value, to_value) -> value
#  @optional id: string | number | function -> string = 0 -- used to share state between values that might cross over between filters, exclusions, and other rules
#  @optional start: number | function -> number = 0
#  @optional duration: number | function -> number = 300
#  @optional delay: number | function -> number = 0
#  @optional ease: EASE | function -> EASE = EASE.OUT_QUART
#  @optional direction: DIRECTION | function -> DIRECTION 
#}
def Animate(**kwargs):
  _custom_id = kwargs.get('id')
  _id = _custom_id if _custom_id != None else get_next_id()
  _interpolate = kwargs.get('interpolate')
  _start = kwargs.get('start')
  _to = kwargs.get('to')
  _duration = kwargs.get('duration')
  _duration = _duration if _duration != None else DEFAULT_DURATION
  _delay = kwargs.get('delay')
  _delay = _delay if _delay != None else DEFAULT_DELAY
  _ease = kwargs.get('ease')
  _ease = _ease if _ease != None else DEFAULT_EASE
  _direction = kwargs.get('direction')
  _direction = _direction if _direction != None else DEFAULT_DIRECTION
  _reset = kwargs.get('reset')
  _reset = _reset if _reset != None else (_direction != DIRECTION.IN_OUT)
  _compare = kwargs.get('compare')
  def animate(style, state):
    id = None
    win = style.win
    if _custom_id != None:
      state = state['custom']
      id = _custom_id(win_state) if callable(_custom_id) else _custom_id
    else:
      state = state['animations']
      id = _id
    if not id in state:
      state[id] = {'value': None}
    state = state[id]
    to = style.resolve(_to, state)
    start = style.resolve(_start, state)
    compare = _compare(to, state.get('last_to')) if callable(_compare) else _compare
    delay = _delay(style, state) if callable(_delay) else _delay
    duration = _duration(style, state) if callable(_duration) else _duration
    direction = _direction(style, state) if callable(_direction) else _direction
    reset = _reset(style, state) if callable(_reset) else _reset
    if compare == False:
      state['change_timestamp'] = GLOBALS.now
    state['last_to'] = to
    time = (GLOBALS.now - (max(win.timestamps['nc'], state.get('change_timestamp') or 0))) * 1000 - (delay or 0)
    # direction in and active means go towards 'to' value from start value
    # direction out and active means go towards start value when the window becomes inactive
    # otherwise the window should be on 'to' value
    # TODO abstract this into a behavior template
    if (direction == DIRECTION.OUT and style._condition == CONDITION.ACTIVE and not style.win.nc) \
      or (direction == DIRECTION.IN and style._condition == CONDITION.INACTIVE and style.win.nc):
        state['value'] = to
        state['start'] = to
        style._animating = False
        return to
    elif (direction == DIRECTION.OUT and style._condition == CONDITION.ACTIVE and style.win.nc) \
      or (direction == DIRECTION.IN and style._condition == CONDITION.INACTIVE and not style.win.nc):
        swp = start
        start = to
        to = swp
    elif (direction == DIRECTION.OUT and style._condition == CONDITION.INACTIVE and not style.win.nc) \
      or (direction == DIRECTION.IN and style._condition == CONDITION.ACTIVE and style.win.nc):
        state['value'] = start
        state['start'] = start
        style._animating = False
        return start
    elif (direction == DIRECTION.OUT and style._condition == CONDITION.INACTIVE and style.win.nc) \
      or (direction == DIRECTION.IN and style._condition == CONDITION.ACTIVE and not style.win.nc):
        pass
    value = state.get('value')
    if value == None:
      state['value'] = start
      state['start'] = start
    if time <= 0:
      if reset == True:
        state['start'] = start
        state['value'] = start
      else:
        state['start'] = state['value']
      style._animating = True
      ANIMATOR.schedule(win)
      return state['start']
    if time >= duration:
      state['value'] = to
      style._animating = False
      return to

    elapsed = time / float(duration)
    elapsed = min(max(_ease(elapsed), 0), 1)
    state['value'] = _interpolate(to, state['start'], elapsed, style, state)
    ANIMATOR.schedule(win)
    style._animating = True
    return state['value']
  return animate
